src/lidar_subscriber.cpp

```diff
@@ -0,0 +1,156 @@
+
+#include <memory>
+#include <cmath>
+#include "rclcpp/rclcpp.hpp"
+
+#include "core-structs.h"
+
+
+#include "sensor_msgs/msg/laser_scan.hpp"
+#include <queue>
+#include <limits>
+#include "visualization_msgs/msg/marker.hpp"
+
+constexpr float LANDMARK_THRESHOLD_m = 0.02;
+constexpr bool testing = true;
+
+bool same_landmark(const Observation2D &first, const Observation2D &second){
+  float r1 = first.range_m;
+  float r2 = second.range_m;
+  float theta1 = first.bearing_rad;
+  float theta2 = second.bearing_rad;
+  float distance = sqrt(pow(r1,2.0)+pow(r2, 2.0)-2*r1*r2*cos(theta1-theta2));
+  if (distance >= LANDMARK_THRESHOLD_m){
+	//std::cout<<distance<<" ";		 
+   }
+  return  distance <= LANDMARK_THRESHOLD_m;
+}
+
+Observation2D merge(Observation2D &existing_landmark, Observation2D &new_measurement, int &count){
+  struct Observation2D average;
+  average.range_m = ((count-1)*existing_landmark.range_m+new_measurement.range_m)/count;
+  if(existing_landmark.bearing_rad>M_PI){
+	  existing_landmark.bearing_rad -=2*M_PI;
+  }
+  if(new_measurement.bearing_rad>M_PI){
+	  new_measurement.bearing_rad-=2*M_PI;
+  }
+  average.bearing_rad = ((count-1)*existing_landmark.bearing_rad+new_measurement.bearing_rad)/count;
+  return average;
+}
+
+Observation2D merge(Observation2D &first, Observation2D &second, const int &count1, const int &count2){
+  //standard weighted average --> between first and last landmark
+  struct Observation2D average;
+  float weight1 = ((float)count1)/((float)(count1+count2));
+  float weight2 = ((float)count2)/((float)(count1+count2));
+  if(first.bearing_rad>M_PI){
+          first.bearing_rad -=2*M_PI;
+  }
+  if(second.bearing_rad>M_PI){
+          second.bearing_rad-=2*M_PI;
+  }
+  average.range_m = first.range_m*weight1+second.range_m*weight2;
+  average.bearing_rad = first.bearing_rad*weight1+(second.bearing_rad)*weight2;
+  //std::cout<<first.bearing_rad<<" " <<second.bearing_rad<<"\n";
+  return average;
+}
+
+class LidarSubscriber : public rclcpp::Node
+{
+  public:
+    LidarSubscriber()
+   : Node("lidar_subscriber")
+    {
+      subscription = this->create_subscription<sensor_msgs::msg::LaserScan>(
+      "scan_filtered", 10, std::bind(&LidarSubscriber::laserscan_to_landmarks, this, std::placeholders::_1));
+      publisher = this->create_publisher<visualization_msgs::msg::Marker>("landmark_visualization",10);
+    }
+  private:
+    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr subscription;
```

Comment:
Following my earlier comment about nesting. `laser_to_landmarks` function should be implemented outside of the class to be more readable. Ideally I should read your class declaration and be instantly clear about the outline of the class, i.e what functions and variables are there. If I want to look up details of implementation I would then look for the function further down in your code. Declaring and implementing the function within your class declaration makes this process slow and confusing, since the reader is getting bogged down by the details even when they are just trying to get a quick sense of what's going on in the class.